<a href="https://colab.research.google.com/github/CodeHunterOfficial/AI_DataMining/blob/main/NLP/Lesson%202.%20%20%D0%92%D0%B5%D0%BA%D1%82%D0%BE%D1%80%D0%BD%D1%8B%D0%B5_%D0%BF%D1%80%D0%B5%D0%B4%D1%81%D1%82%D0%B0%D0%B2%D0%BB%D0%B5%D0%BD%D0%B8%D1%8F_%D1%81%D0%BB%D0%BE%D0%B2_(Word2Vec%2CGloVe%2CELMo%2CSkip_Gram)/2_0_3_WordPiece.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#WordPiece

### Введение в WordPiece

**WordPiece** — это алгоритм сегментации текста, разработанный для улучшения представления и обработки текстов на естественном языке. Изначально WordPiece был создан для улучшения качества распознавания речи и машинного перевода, но со временем алгоритм нашел широкое применение в задачах обработки естественного языка, особенно в языковых моделях, таких как BERT от Google.

WordPiece относится к классу методов токенизации, которые разбивают текст на токены — минимальные значимые единицы текста, используемые для анализа и обработки текста. Основная идея алгоритма WordPiece заключается в том, чтобы разбиение текста происходило не только на уровне слов, но и на уровне подслов (subwords). Это позволяет алгоритму справляться с редкими или неизвестными словами, разбивая их на более мелкие компоненты, которые уже были обучены и содержат полезную информацию.

### Задачи и проблемы, решаемые WordPiece

1. **Решение проблемы неизвестных слов**: В традиционных подходах токенизации каждое уникальное слово записывается в словарь, и если слово встречается впервые, его классифицируют как "неизвестное" (UNK-token). Это вызывает сложности, особенно при обработке слов, которых не было в обучающем наборе данных. WordPiece решает эту проблему, разбивая редкие или неизвестные слова на известные подслова, которые уже присутствуют в словаре модели.

2. **Сокращение размера словаря**: При работе с большими наборами данных на естественном языке словарь может достигать огромных размеров, что увеличивает требования к памяти и время обработки. WordPiece снижает количество уникальных токенов, записывая их как комбинации подслов, что позволяет использовать меньше памяти и обрабатывать текст быстрее.

3. **Улучшение качества моделирования морфологически сложных языков**: В некоторых языках (например, в русском, финском или турецком) слова могут содержать множество префиксов, суффиксов и других морфем, которые значительно изменяют значение и форму слова. WordPiece учитывает морфемные структуры и позволяет эффективнее моделировать подобные языки, делая модель более точной.

### Алгоритм WordPiece: пошаговое объяснение

Процесс обучения WordPiece включает в себя несколько основных шагов:

#### 1. Сбор исходного текста
Для создания словаря WordPiece необходимо начать с большого корпуса текстов, на основе которого будут определяться подслова. Чем больше и разнообразнее корпус, тем лучше алгоритм сможет выделить устойчивые и полезные подслова.

#### 2. Начальная сегментация символов
На первом шаге текст разбивается на отдельные символы. Каждое слово рассматривается как последовательность символов, и каждый символ является потенциальным подсловом. Например, слово "книга" можно представить как [к, н, и, г, а].

#### 3. Определение частотных биграмм
WordPiece использует частотные биграммы (пары символов) для объединения символов в подслова. Изначально каждому символу присваивается его частотность в корпусе, а затем рассчитываются частотности биграмм. Чем чаще встречается биграмма, тем вероятнее, что она станет подсловом.

Пример:
- Если в корпусе часто встречаются биграммы "кн" и "иг", WordPiece может объединить их, чтобы получить подслово "книг".

#### 4. Итеративное объединение биграмм
На каждой итерации алгоритм объединяет наиболее частые биграммы в подслова. Этот процесс продолжается до тех пор, пока не будет достигнуто заданное количество подслов или пока не исчерпаются все возможные биграммы.

Итак, после нескольких итераций "книг" и "а" могут быть объединены в подслово "книга". Таким образом, реже встречающиеся слова могут быть разбиты на несколько подслов, а часто встречающиеся — представлены как один токен.

#### 5. Ограничение размера словаря
С помощью WordPiece можно задать максимальный размер словаря, например, 30 000 подслов. Ограничение размера словаря позволяет контролировать объем памяти и вычислительных ресурсов, необходимых для хранения и обработки модели.

#### 6. Запись подслов в словарь
После выполнения всех итераций все подслова записываются в словарь, и алгоритм готов к использованию. При обработке нового текста WordPiece разбивает слова на подслова, опираясь на этот словарь.

### Пример работы WordPiece

Рассмотрим, как WordPiece работает на примере слова "недоступность".

1. В начале слова сегментируются на символы: "н", "е", "д", "о", "с", "т", "у", "п", "н", "о", "с", "т", "ь".
2. Затем WordPiece начинает объединять символы в биграммы, такие как "не", "до", "ст", "уп".
3. После нескольких итераций алгоритм может прийти к следующему набору подслов: ["не", "доступ", "ность"].
4. Теперь вместо использования 13 отдельных символов мы представляем слово как комбинацию из трех подслов.

Такой подход позволяет WordPiece эффективно обрабатывать не только распространенные слова, но и редкие или даже придуманные слова, которые, вероятно, состоят из знакомых модели подслов.

### Применение WordPiece в NLP моделях

WordPiece используется в таких популярных языковых моделях, как BERT, DistilBERT и многих других моделях глубокого обучения. Разберем несколько примеров, как токенизация WordPiece способствует их эффективности:

1. **BERT**: BERT (Bidirectional Encoder Representations from Transformers) использует WordPiece для создания токенов входного текста. Благодаря этому модель может понимать, что слова вроде "счастливый", "счастливы" и "счастье" связаны между собой, поскольку содержат общие подслова. Это позволяет модели генерировать более богатые представления слов и улучшает качество предсказаний.

2. **Машинный перевод**: В задачах перевода WordPiece позволяет точно токенизировать редко встречающиеся слова, что улучшает понимание текста, точность перевода и качество общего результата.

3. **Распознавание речи**: В системах ASR (Automatic Speech Recognition) WordPiece помогает точно интерпретировать слова, даже если они произносятся неточно или отсутствуют в словаре. Это улучшает качество распознавания речи для редких слов или слов на нескольких языках.

### Преимущества и недостатки WordPiece

#### Преимущества:
1. **Гибкость в работе с неизвестными словами**: WordPiece позволяет разбирать неизвестные слова на подслова, что делает модель более устойчивой к незнакомым или редким словам.
2. **Снижение объема словаря**: Благодаря комбинированию символов в подслова, WordPiece позволяет уменьшить размер словаря, что экономит память и ускоряет обучение.
3. **Улучшение работы с морфологически сложными языками**: WordPiece эффективно разбивает слова на морфемы, что позволяет моделям лучше понимать языки с богатой морфологией.

#### Недостатки:
1. **Неоптимальность для простых языков**: Для языков, где структура слов проста и изменяется редко, WordPiece может быть излишним и добавлять ненужную сложность.
2. **Медленная токенизация**: Поскольку WordPiece использует итеративное объединение биграмм, процесс токенизации может занимать больше времени, чем у простых методов, например, у Byte-Pair Encoding (BPE).
3. **Фиксированный словарь**: Словарь, созданный WordPiece, фиксирован на этапе обучения и не может адаптироваться к новым словам без повторного обучения модели.

### Сравнение WordPiece с другими методами токенизации

- **WordPiece vs. Byte-Pair Encoding (BPE)**: Оба метода используют идею подслов и объединяют символы в подслова на основе частотности биграмм. Однако BPE часто требует меньше итераций для достижения аналогичного результата и может работать быстрее.
  
- **WordPiece vs. Unigram**: Unigram сегментирует слова, минимизируя вероятность всей последовательности. WordPiece, напротив, итеративно ищет самые частотные биграммы, что делает его более гибким для подбора подслов, хотя и более ресурсоемким.

- **WordPiece vs. Простая токенизация по словам**: Традиционная токенизация по словам выделяет каждое слово как отдельный токен, но не может обработать неизвестные слова. WordPiece решает эту проблему, делая токенизацию более надежной, но за счет повышения сложности модели.

Таким образом, алгоритм WordPiece стал важным инструментом в области обработки естественного языка. Благодаря способности эффективно разбирать слова на подслова, WordPiece не только облегчает обработку редких или неизвестных слов, но и улучшает качество понимания текста и результатов моделирования.

Реализация WordPiece на Python
Теперь создадим простую реализацию алгоритма WordPiece на Python. Она включает в себя сегментацию символов, подсчет частотности биграмм и объединение биграмм.

In [ ]:
from collections import Counter

# Исходный текст
corpus = ["недоступность", "компьютера", "мешает", "пользователям"]

# Функция для разбиения слов на символы
def tokenize_characters(word):
    return list(word)

# Подсчет частотности биграмм
def get_bigram_frequencies(words):
    bigram_freq = Counter()
    for word in words:
        for i in range(len(word) - 1):
            bigram = (word[i], word[i + 1])
            bigram_freq[bigram] += 1
    return bigram_freq

# Функция для объединения наиболее частотной биграммы
def merge_most_frequent_bigram(words, most_frequent_bigram):
    new_words = []
    for word in words:
        new_word = []
        i = 0
        while i < len(word):
            if i < len(word) - 1 and (word[i], word[i + 1]) == most_frequent_bigram:
                # Объединяем биграмму
                new_word.append(word[i] + word[i + 1])
                i += 2  # Пропускаем следующий символ
            else:
                new_word.append(word[i])
                i += 1
        new_words.append(new_word)
    return new_words

# Основная функция WordPiece
def train_wordpiece(corpus, vocab_size=30):
    words = [tokenize_characters(word) for word in corpus]
    vocab = set(sum(words, []))  # начальный словарь с символами

    # Пока размер словаря меньше нужного, объединяем биграммы
    while len(vocab) < vocab_size:
        bigram_freq = get_bigram_frequencies(words)

        # Если биграммы закончились, завершаем процесс
        if not bigram_freq:
            break

        # Находим наиболее частую биграмму
        most_frequent_bigram = max(bigram_freq, key=bigram_freq.get)

        # Объединяем эту биграмму в словах
        words = merge_most_frequent_bigram(words, most_frequent_bigram)

        # Обновляем словарь, добавляя объединенную биграмму как новое подслово
        vocab.add("".join(most_frequent_bigram))

    return vocab, words

# Запускаем обучение WordPiece
vocab, tokenized_words = train_wordpiece(corpus, vocab_size=20)

print("Словарь подслов:")
print(vocab)
print("\nТокенизированные слова:")
for word in tokenized_words:
    print("".join(word))

Словарь подслов:
{'н', 'а', 'у', 'в', 'ю', 'ь', 'л', 'к', 'о', 'я', 'е', 'ос', 'р', 'т', 'ш', 'з', 'м', 'с', 'д', 'п'}

Токенизированные слова:
недоступность
компьютера
мешает
пользователям



### Пояснение к коду

1. **Разбиение слов на символы**: Функция `tokenize_characters` разбивает каждое слово на отдельные символы.
   
2. **Подсчет частотности биграмм**: Функция `get_bigram_frequencies` считает частотность биграмм для всех слов в корпусе. Это нужно для нахождения наиболее частотных комбинаций символов, которые мы будем объединять в подслова.

3. **Объединение биграмм**: В функции `merge_most_frequent_bigram` происходит объединение биграммы с самой высокой частотностью. Если биграмма встречается в слове, она объединяется в один токен, и слово обновляется.

4. **Обучение WordPiece**: Основная функция `train_wordpiece` выполняет итерации по всему процессу, пока не достигнут заданный размер словаря подслов `vocab_size`. В итоге возвращается словарь подслов и список токенизированных слов.


В Python существуют готовые библиотеки, которые реализуют алгоритм WordPiece и позволяют использовать его для токенизации текста. Одной из таких популярных библиотек является `transformers` от компании Hugging Face. Эта библиотека включает в себя различные модели NLP, такие как BERT, которые используют WordPiece в качестве алгоритма токенизации.

Давайте посмотрим, как можно использовать токенизацию WordPiece с помощью библиотеки `transformers`.

### Установка библиотеки `transformers`

Если `transformers` еще не установлена, её можно установить командой:


In [ ]:
!pip install transformers


### Использование WordPiece с библиотекой `transformers`

Для демонстрации работы WordPiece возьмем модель BERT, которая использует данный алгоритм токенизации. Токенайзер модели BERT построен на базе WordPiece и автоматически выполняет все необходимые шаги для токенизации текста.


In [ ]:
from transformers import BertTokenizer

# Загружаем токенайзер, основанный на WordPiece
tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")

# Пример текста для токенизации
text = "Tokenization using WordPiece in natural language processing."

# Токенизируем текст
tokens = tokenizer.tokenize(text)

# Отображаем токены
print("Токены:")
print(tokens)

# Преобразуем токены в идентификаторы (индексы) для подачи в модель
token_ids = tokenizer.convert_tokens_to_ids(tokens)

print("\nИдентификаторы токенов:")
print(token_ids)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

Токены:
['token', '##ization', 'using', 'word', '##piece', 'in', 'natural', 'language', 'processing', '.']

Идентификаторы токенов:
[19204, 3989, 2478, 2773, 11198, 1999, 3019, 2653, 6364, 1012]



### Объяснение работы кода

1. **Загрузка токенайзера**: Мы загружаем предобученный токенайзер `BertTokenizer` из библиотеки `transformers`. Это автоматически загружает модель, которая использует алгоритм WordPiece для сегментации текста.

2. **Токенизация текста**: С помощью метода `tokenizer.tokenize()` мы разбиваем текст на подслова. WordPiece выполняет сегментацию на уровне подслов, так что редкие или сложные слова будут разделены на более мелкие части.

3. **Преобразование токенов в индексы**: Метод `convert_tokens_to_ids()` переводит токены в числовые индексы. Эти индексы соответствуют позициям токенов в словаре модели и могут быть поданы на вход модели BERT для дальнейшей обработки.



- Здесь `##` перед подсловом указывает, что это часть слова. Например, "token" и "##ization" — это подслова, составляющие слово "tokenization".
- Числовые индексы (идентификаторы токенов) — это представление токенов, которое модель использует для дальнейшего анализа.



### Дополнительные возможности токенайзера

Токенайзер из библиотеки `transformers` также предоставляет возможность добавить специальные токены, такие как `[CLS]` и `[SEP]`, которые полезны при работе с моделями вроде BERT. Пример использования этих токенов:


In [ ]:
# Токенизация текста с добавлением специальных токенов
encoded_input = tokenizer(text, return_tensors="pt", padding=True, truncation=True)

print("\nЗакодированные данные с добавлением специальных токенов:")
print(encoded_input)


Закодированные данные с добавлением специальных токенов:
{'input_ids': tensor([[  101, 19204,  3989,  2478,  2773, 11198,  1999,  3019,  2653,  6364,
          1012,   102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}


Давайте разберем, что означает каждый из полей в закодированных данных, которые вернул токенайзер:

### 1. `input_ids`
```python
'input_ids': tensor([[  101, 19204,  3989,  2478,  2773, 11198,  1999,  3019,  2653,  6364, 1012,   102]])
```

Это последовательность числовых идентификаторов, которые соответствуют токенам текста. Эти идентификаторы относятся к уникальным токенам из словаря модели.

- `101`: Специальный токен `[CLS]`, добавленный в начало. Он используется как индикатор начала текста и особенно важен для задач классификации.
- `102`: Специальный токен `[SEP]`, добавленный в конец текста. Он используется для обозначения конца текста или разделения предложений.
- Другие числа представляют токены, которые были выделены с использованием алгоритма WordPiece.

Примерный перевод каждого из токенов (по их исходному тексту):

```
[CLS] token ##ization using word ##piece in natural language processing . [SEP]
```

Токены с префиксом `##` являются подсловами, присоединенными к предыдущим частям слов (например, "##ization" добавляется к "token", образуя "tokenization").

### 2. `token_type_ids`
```python
'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])
```

`token_type_ids` указывает типы токенов и используется для определения предложений при входе модели. Обычно этот массив применяется в задачах, связанных с парными предложениями (например, при сопоставлении вопросов и ответов), чтобы отличить одно предложение от другого. В данном случае все значения равны `0`, так как текст состоит только из одного предложения.

### 3. `attention_mask`
```python
'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])
```

`attention_mask` указывает, какие токены следует учитывать модели при обработке, где `1` обозначает значимый токен, а `0` — токен-заполнитель (padding). Здесь все значения равны `1`, что означает, что все токены значимы и участвуют в анализе текста, так как здесь нет дополнительных токенов-заполнителей.

### Общий смысл

Эти данные представляют собой полный вход для модели, подготовленный токенайзером. Его можно передать непосредственно в модель BERT для дальнейшей обработки, такой как извлечение признаков текста или решение конкретной задачи, например, классификации текста, анализа настроения и т. д.